File: pantompkins.m
close all
clear all
load data.mat

%MISSING SEARCH BACK

for j=1:1:10

fs = 500; %samples/sec
ECG = x_train(j,:,1);

%% PREPROCESSING

%Low pass filter (need to use hexadecimal and account for FPGA accuracy)
f1=5;                                                                      % cuttoff low frequency to get rid of baseline wander
f2=15;                                                                     % cuttoff frequency to discard high frequency noise
Wn=[f1 f2]*2/fs;                                                           % cutt off based on fs
N = 6;                                                                     % order of 3 less processing
[a,b] = butter(N,Wn);  
pre1 = filter(a,b,ECG);

%Derivative filter
pre2 = filter([2 1 0 -1 -2],[1],pre1)/8;

%Squaring function
pre3 = pre2.^2;

%Moving window integration
window_length = 0.150; %s, PAN TOMPKINS paper chooses N so that the length of window is 150ms
N = window_length*fs;
pre4 = filter(ones(N,1),[1],pre3);

% figure;
% subplot(3,2,1);
% plot(ECG);
% subplot(3,2,2);
% plot(pre1);
% subplot(3,2,3);
% plot(pre2);
% subplot(3,2,4);
% plot(pre3);
% subplot(3,2,5);
% plot(pre4);

%% BEAT DETECTION
filtered = pre4;
increasing = 1;
out = zeros(length(filtered),1);
SPK = zeros(length(filtered),1);
NPK = zeros(length(filtered),1);
THR1 = zeros(length(filtered),1);
THR2 = zeros(length(filtered),1);
refractory_period = 0.2*fs;
count = 0;
RR = [];
max = 0;
above_THR1 = 0;
peak_i = 0;
peak = 0;

for i=2:1:length(filtered)
    if i<2*fs
        if filtered(i) > max 
            max = filtered(i);
        end
        SPK(i)=max; %ORIGINAL DIVIDES BY 4
        NPK(i) = NPK(i-1) + 0.5*filtered(i)/(2*fs);
    else
        if length(RR) > 8
           RR_ave = mean(RR(end-7:end));
        else
           RR_ave = mean(RR);
        end

        if (filtered(i-2)<filtered(i-1)) && (filtered(i)<filtered(i-1))
            if (filtered(i-1)>THR1(i-1))
                if count<=refractory_period
                    if peak < filtered(i-1)
                        if peak_i ~= 0 out(peak_i) = 0; end;                        
                        SPK(i) = 0.125*filtered(i-1) + 0.875*SPK(i-1);
                        RR = [RR; count];
                        out(i-1) = filtered(i-1);
                        peak_i = i-1;
                        peak = filtered(i-1);
                        count = 0;
                    else
                        NPK(i) = 0.125*filtered(i-1) + 0.875*NPK(i-1);
                    end
                else
                    SPK(i) = 0.125*filtered(i-1) + 0.875*SPK(i-1);
                    RR = [RR; count];
                    out(i-1) = filtered(i-1);
                    peak_i = i-1;
                    peak = filtered(i-1);
                    count = 0;
                end
            else
                NPK(i) = 0.125*filtered(i-1) + 0.875*NPK(i-1);
            end
        end
        if SPK(i)==0 
            SPK(i)=SPK(i-1);
        end
        if NPK(i)==0 
            NPK(i)=NPK(i-1);
        end
        THR1(i) = NPK(i) + 0.25*(SPK(i)-NPK(i));
        THR2(i) = THR1(i)/8;
        count = count + 1;
    end
end

figure;
hold on;
plot(filtered);
stem(out);
plot(THR1,'Color','red');
plot(THR2,'Color','green');
hold off;

%figure;
%pan_tompkin(ECG, fs, 1);

end

File: pan-tompkins.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import signal

x_train = np.load('x_train.npy', allow_pickle=True)
x_test = np.load('x_test.npy', allow_pickle=True)
y_train = np.load('y_train.npy', allow_pickle=True)
y_test = np.load('y_test.npy', allow_pickle=True)

####NEEED TO CHANGE FILTER COEFFICIENTS BECAUSE OF DIFFERENT SAMPLLING FREQUENCY

f_s = 100 #samples/sec

#YOU CAN CHOOSE ANY INPUT FROM THE DATABASE
ECG = x_train[0,:,0]

#Low pass filter (need to account for FPGA accuracy)
h = np.array([-15, -13, 7, 38, 66, 38, 7, -13, -15])
preprocessed1 = np.convolve(ECG, h)

#Derivative filter
preprocessed2 = signal.lfilter(np.array([2, 1, 0, -1, -2]),[1],preprocessed1)/8

#Squaring function
preprocessed3 = np.square(preprocessed2)

#Moving window integration
window_length = 0.150 #s, PAN TOMPKINS paper chooses N so that the length of window is 150ms
N = int(window_length*f_s)
preprocessed4 = signal.lfilter(np.ones(N),[1],preprocessed3)

fig, axs = plt.subplots(4,1)
axs[0].plot(ECG)
axs[1].plot(preprocessed1)
axs[2].plot(preprocessed3)
axs[3].plot(preprocessed4)
fig.suptitle('Different steps of the preprocessing')

## BEAT DETECTION
state = 0
previous_sample = 0
increasing = 1
initial_period = 2*f_s
THR1 = 0
THR2 = 0
NPK = 0
SPK = 0
out = np.zeros(preprocessed4.size)
refractory_period = int(0.2*f_s)
count = 0
previous_count = 0
RR = np.empty(shape=(0,0))

for i in range(preprocessed4.size):
    if RR.size > 8:
        RR_ave = np.mean(RR[-8:])
    else:
        RR_ave = np.mean(RR)
    if (increasing == 1) and (preprocessed4[i]<preprocessed4[i-1]):
        if ((preprocessed4[i-1]>THR1) or ((preprocessed4[i-1]>THR2) and (count>=1.16*RR_ave))) and (count>refractory_period):
            if i<200:
                SPK = preprocessed4[i-1]  
            else:
                SPK = 0.125*preprocessed4[i-1] + 0.875*SPK
                RR = np.append(RR, count)
            out[i-1] = 1
            count = 0
        else:
            if i<200:
                NPK = preprocessed4[i-1]  
            else:
                NPK = 0.125*preprocessed4[i-1] + 0.875*NPK
        increasing = 0
    if (increasing == 0) and (preprocessed4[i]>=preprocessed4[i-1]):
        increasing = 1
    THR1 = NPK + 0.25*(SPK-NPK)
    THR2 = THR1/8
    count = count + 1

fig2, ax2 = plt.subplots(1,1)
ax2.plot(out)
fig2.suptitle('Final result of pan-tompkins')
plt.show()
